pkg/steps/artifacts_test.go

```diff
@@ -712,3 +719,108 @@ func TestAddPodUtils(t *testing.T) {
 	}
 	testhelper.CompareWithFixture(t, base)
 }
+
+func aPod() *coreapi.Pod {
+	return &coreapi.Pod{
+		ObjectMeta: meta.ObjectMeta{
+			Name:      "p",
+			Namespace: "ns",
+		},
+	}
+}
+
+func TestWaitForConditionOnObject(t *testing.T) {
+	containerName := "c"
+	podName := "p"
+	ns := "ns"
+
+	evaluateFunc := func(obj runtime.Object) (bool, error) {
+		switch pod := obj.(type) {
+		case *coreapi.Pod:
+			for _, container := range pod.Status.ContainerStatuses {
+				if container.Name == containerName {
+					if container.State.Running != nil || container.State.Terminated != nil {
+						return true, nil
+					}
+					break
+				}
+			}
+		default:
+			return false, fmt.Errorf("pod/%v ns/%v got an event that did not contain a pod: %v", podName, ns, obj)
+		}
+		return false, nil
+	}
+
+	testCases := []struct {
+		name          string
+		expected      error
+		client        watchingclient.Client
+		identifier    ctrlruntimeclient.ObjectKey
+		list          ctrlruntimeclient.ObjectList
+		into          ctrlruntimeclient.Object
+		evaluate      evaluator
+		timeout       time.Duration
+		containerName string
+		objectFunc    func(client ctrlruntimeclient.Client)
+	}{
+		{
+			name:       "happy path: pod",
+			client:     fakectrlruntimeclient.NewFakeClient(aPod()),
+			identifier: ctrlruntimeclient.ObjectKey{Name: podName, Namespace: ns},
+			list:       &coreapi.PodList{},
+			into:       &coreapi.Pod{},
+			timeout:    time.Second,
+			evaluate:   evaluateFunc,
+			objectFunc: func(client ctrlruntimeclient.Client) {
+				_ = wait.Poll(100*time.Millisecond, time.Second,
+					func() (bool, error) {
+						ctx := context.TODO()
+						pod := &coreapi.Pod{}
+						if err := client.Get(ctx, ctrlruntimeclient.ObjectKey{Name: podName, Namespace: ns}, pod); err != nil {
+							if kerrors.IsNotFound(err) {
+								return false, nil
+							}
+							return false, err
+						}
+						pod.Status.ContainerStatuses = append(pod.Status.ContainerStatuses, coreapi.ContainerStatus{
+							Name: "c",
+							State: coreapi.ContainerState{
+								Running: &coreapi.ContainerStateRunning{},
+							},
+						})
+						if err := client.Update(ctx, pod); err != nil {
+							return false, err
+						}
+						return true, nil
+					},
+				)
+			},
+		},
+		{
+			name:       "timeout",
+			client:     fakectrlruntimeclient.NewFakeClient(aPod()),
+			identifier: ctrlruntimeclient.ObjectKey{Name: podName, Namespace: ns},
+			list:       &coreapi.PodList{},
+			into:       &coreapi.Pod{},
+			timeout:    time.Second,
+			evaluate:   evaluateFunc,
+			objectFunc: func(client ctrlruntimeclient.Client) {},
+			expected:   fmt.Errorf("timed out waiting for the condition"),
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			var wg sync.WaitGroup
+			wg.Add(1)
```

Comment:
A `waitGroup` whose `Wait` is never called is useless. I don't see why we need synchronization here though

Reply:
I was jumping between implementation ideas.